webapi/Skills/ChatSkills/ChatSkill.cs

```diff
@@ -337,43 +336,165 @@ public async Task<string> ExtractAudienceAsync(SKContext context, CancellationTo
+        context.Variables.Update(chatMessage.Content);
+
+        if (chatMessage.TokenUsage != null)
+        {
+            context.Variables.Set("tokenUsage", JsonSerializer.Serialize(chatMessage.TokenUsage));
+        }
+        else
+        {
+            this._logger.LogWarning("ChatSkill.ChatAsync token usage unknown. Ensure token management has been implemented correctly.");
+        }
+
+        return context;
+    }
+
+    /// <summary>
+    /// This is the entry point for handling a plan, whether the user approves, cancels, or re-runs it.
+    /// </summary>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    [SKFunction, Description("Process a plan")]
+    public async Task<SKContext> ProcessPlanAsync(
+        [Description("The new message")] string message,
+        [Description("Unique and persistent identifier for the user")] string userId,
+        [Description("Name of the user")] string userName,
+        [Description("Unique and persistent identifier for the chat")] string chatId,
+        [Description("Proposed plan object"), DefaultValue(null), SKName("proposedPlan")] string? planJson,
+        SKContext context,
+        CancellationToken cancellationToken = default)
+    {
+        // Ensure that plan exists in ask's context variables.
+        // If a plan was returned at this point, that means it is a
+        //      1. Proposed plan that was approved or cancelled, or
+        //      2. Saved plan being re-run.
+        ProposedPlan? deserializedPlan;
+        if (string.IsNullOrWhiteSpace(planJson) || (deserializedPlan = JsonSerializer.Deserialize<ProposedPlan>(planJson)) == null)
+        {
+            throw new ArgumentException("Plan does not exist in request context. Ensure that a valid plan is saved in the ask's variables");
+        }
+
+        // Ensure plan is actionable by supported planners.
+        if (!(deserializedPlan.Type == PlanType.Action || deserializedPlan.Type == PlanType.Sequential))
+        {
+            throw new ArgumentException($"Plan of type {deserializedPlan.Type} cannot be executed. Only Action or Sequential plans can be re-run at this time.");
+        }
+
+        // Set the system description in the prompt options.
+        await this.SetSystemDescriptionAsync(chatId, cancellationToken);
+
+        // Clone the context to avoid modifying the original context variables.
+        var chatContext = context.Clone();
+        chatContext.Variables.Set("knowledgeCutoff", this._promptOptions.KnowledgeCutoffDate);
+
+        // Save this new message to memory such that subsequent chat responses can use it
+        await this.UpdateBotResponseStatusOnClientAsync(chatId, "Saving user message to chat history", cancellationToken);
+        var newUserMessage = await this.SaveNewMessageAsync(message, userId, userName, chatId, ChatMessage.ChatMessageType.Message.ToString(), cancellationToken);
```

Comment:
looks like we could have a helper fn for this logic and use it in both routes?

Reply:
Can leave for another PR if we want to refactor it, but I figured I'd leave it like this for now because it allows more flexibility in manipulating the chatContext as well as fine tuning the system description, if needed.